hotham/src/components/ui_panel.rs

```diff
@@ -0,0 +1,170 @@
+use ash::vk::{self};
+use egui::emath::vec2;
+use egui::epaint::Vertex as EguiVertex;
+use egui::CtxRef;
+use hecs::{Entity, World};
+use nalgebra::{Vector2, Vector3};
+use rapier3d::prelude::{ColliderBuilder, InteractionGroups};
+
+const BUFFER_SIZE: usize = 1024;
+
+use crate::buffer::Buffer;
+use crate::components::Panel;
+use crate::resources::gui_context::SCALE_FACTOR;
+use crate::resources::physics_context::PANEL_COLLISION_GROUP;
+use crate::resources::{GuiContext, PhysicsContext};
+use crate::resources::{RenderContext, VulkanContext};
+
+use super::{Collider, Transform, TransformMatrix, Visible};
+
+/// A component added to an entity to display a 2D "panel" in space
+/// Used by `panels_system`
+#[derive(Clone)]
+pub struct UIPanel {
+    /// The text to be displayed
+    pub text: String,
+    /// Framebuffer this panel will be written to
+    pub framebuffer: vk::Framebuffer,
+    /// Vertices of the panel
+    pub vertex_buffer: Buffer<EguiVertex>,
+    /// Indices of the panel
+    pub index_buffer: Buffer<u32>,
+    /// Reference to egui context
+    pub egui_context: CtxRef,
+    /// The raw input for this panel this frame
+    pub raw_input: egui::RawInput,
+    /// A list of buttons in this panel
+    pub buttons: Vec<UIPanelButton>,
+}
+
+/// A button for a panel
+#[derive(Debug, Clone)]
+pub struct UIPanelButton {
+    /// Text to be displayed
+    pub text: String,
+    /// Was this button clicked?
+    pub clicked_this_frame: bool,
+}
+
+impl UIPanelButton {
+    /// Convenience function to create a new panel button
+    pub fn new(text: &str) -> Self {
+        UIPanelButton {
+            text: text.to_string(),
+            clicked_this_frame: false,
+        }
+    }
+}
+
+/// Convenience function to create a panel and add it to a World
+#[cfg_attr(feature = "cargo-clippy", allow(clippy::too_many_arguments))]
+pub fn add_panel_to_world(
```

Comment:
nit: This should now be renamed to `add_ui_panel_to_world`

Reply:
Done.